src/NUnitFramework/framework/Internal/TestExecutionContext.cs

```diff
@@ -428,6 +428,12 @@ public IPrincipal CurrentPrincipal
         /// </summary>
         public bool IsSingleThreaded { get; set; }
 
+        /// <summary>
+        /// The number of times the current test has been scheduled for execution.
+        /// Currently only being executed in a test using the <see cref="RetryAttribute"/>
+        /// </summary>
+        public int CurrentRepeatCount { get; set; }
```

Comment:
I feel less than comfortable making the setter public. I'd prefer `{ get; private set; }` with a public `void IncrementCurrentRepeatCount()` method and an *internal* `void ResetCurrentRepeatCount()` method, if we need one. Requesting changes just to discuss this. Thoughts?

Reply:
This would also semantically give us the opportunity in the future to do other work during `IncrementCurrentRepeatCount()` such as raising an event for runners to react to.

Reply:
But we don't do other work and aren't even considering it, AFAIK. If we decided to issue an event on retry (which I feel is far-fetched anyway) we would hardly do it as a side effect of incrementing a counter! This is just one of many property setters in `TestExecutionContext`. Shouldn't we present a consistent interface to those who use the class?

Reply:
Even if we are deciding now that we aren't going to raise an event, I still want the user code to be forced to use an API indicating that it is repeating the test rather (semantic) than an API which directly affects the counter (implementation). In that regard, `IncrementCurrentRepeatCount` (which I suggested to illustrate the direction I wanted to move in) is a poor name. I'd prefer `BeginTestRepeat()`.

I do agree about consistency so that it doesn't get lost in the other properties, but in that case I really dislike the fact that it is a loose counter. What about a domain object?

```cs
public class TestExecutionContext
{
    public TestRepeatTracker CurrentRepeat { get; }
}
public sealed class TestRepeatTracker
{
    public int Number { get; }
    public void StartNew();
}
```

User attributes call TestExecutionContext.CurrentRepeat.StartNew() and user tests log TestExecutionContext.CurrentRepeat.Number. Resetting is done by internally setting TestExecutionContext.CurrentRepeat` to a new instance.

To me this solution more closely mirrors the problem space. Offering a loose counter variable also solves the problem but it's a leakier abstraction with more implied, unexpressed expectations. How should code react to the possibility that unrelated code might set `-1` or count backwards? I'm afraid of subtle misunderstandings of the unspoken convention, such as what 0 or 1 mean, so I'm trying to encode the expectations into the API.
Once we ship a loose variable and people start depending on being able to set it without following a particular convention, we'll end up being stuck with it even if we want to add a feature on top of it in the future unless we break people.

Now that's my point of view but I don't hold it strongly and don't want to impose if no one is enamored of domain objects to maintain.

Reply:
@jnm2 I think you might be misinterpreting what happens here - or if not it won't hurt for me to spell it out...

The __attribute__ doesn't itself increment anything. What it does is provide a __command__ that retries the test. The framework simply checks to see if the attribute wants to provide a command and then uses it. Some attributes instantiate an external command while others define it as a nested class. 

That command becomes a temporary part of the execution term and it actually executes the test. So, in other words, it doesn't retry the test as a side effect of incrementing. Rather, the increment is the side effect and isn't used by NUnit in any other way.

To put it another way, the command doesn't __tell__ NUnit to retry the test, it simply __does__ it. As far as NUnit is concerned, there is only one test execution. and the core parts of the framework don't even know the test was retried. So there is no API that causes a test to be repeated.,, we're just recording some new information.

As a minor detail... I'm not understanding why you would ever want to reset the value. That would seem to be an error if it has once been incremented. I could imagine that you might want to allow it to __only__ be incremented rather than set, similar to how we handle assert counts.

Reply:
I was counting the command code as attribute code, but I'll be more precise. I do understand the picture. I'm imagining the user attribute's command's code doing something vaugely like this:

```cs
for (var i = 0; i < max; i++)
{
    innerCommand.Execute(); // Actually perform the execution/repeat
    TestExecutionContext.CurrentContext.CurrentRepeat.StartNew(); // Report new repeat
}
```

I prefer that to `TestExecutionContext.CurrentContext.CurrentRepeatCount = i + 1;`.

You're right, there's no reason to reset since the context is no reused between tests. I wasn't sure.

Reply:
Speaking of assert counts, I wonder if it would be interesting to keep this count on the current result rather than the context? The current result could link to a list of results from the previous repeats. Then the .Count of that collection would be the source of truth for the repeat count property which users could log.

Reply:
@jnm2 Regarding the code for a command... I would imagine that the increment or start new would have to take place before executing the inner command. Otherwise, the test would not be able to access the value from the TestContext. But otherwise... yes that makes sense if we want NUnit to know that repeats are being performed. But what would NUnit do with that info?

Regarding assert counts, there's a chicken and egg problem here. At any point where the user code can access the count, then the count is still unfinished. That is, the user can issue an assert at any of those points. Keeping it in the context was intended to signal that it's provisional.

Right now we throw away results from earlier repetitions.

Reply:
In the initial execution of the inner command, no repeats have begun yet the way I was thinking. The second time the inner commend is executed is the first time it is repeated. But I can adopt another terminology.

What would NUnit do with that info? At the least, NUnit could increment the repeat count property. The motivation of having the API step back and express *intent* rather than *implementation* (~like turn signal rather than changing lanes) was more for the benefit of the API consumers than for NUnit. When semantics are only represented as implementation details, the API is more error-prone for the users.
I went on to imagine how the API shape could potentially yield NUnit richer opportunities but even without that I'd still like the API shape.

Now I'm wondering if we can fit it together with test results and make it possible to see all results when diagnosing a problem. It would be cool not to throw it away. Assuming we stop throwing it away, there is where our repeat count would most naturally come from. I'm thinking that's better than what I proposed first.